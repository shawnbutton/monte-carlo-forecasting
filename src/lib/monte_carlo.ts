const runForWeek = (samples: number[], weeks: number) => {
	const numSamples = samples.length;

	const sims: number[] = [];
	for (let i = 0; i < 10000; i++) {

		let stories = 0;

		for (let week = 0; week < weeks; week++) {
			const index = Math.floor(Math.random() * numSamples);
			stories += samples[index];
		}

		sims.push(stories);
	}

	const sorted = sims
		.sort((n1, n2) => n1 - n2)
		.reverse();

	const results: number[] = [];

	for (let i = 0; i < 99; i++) {
		results[i] = sorted[i * 100];
	}

	return results;

};

const runSimulations = (samples: number[]) => {
	const numSamples = samples.length;

	const sims: number[] = [];

	for (let i = 0; i < 100000; i++) {
		const index = Math.floor(Math.random() * numSamples);
		sims.push(samples[index]);
	}

	const sortedSims = sims
		.sort((n1, n2) => n1 - n2);

	console.log(sortedSims);

};

const runForRangeOfWeeks = (samples: number[], weeks: number) => {
	const allSims = []
	for (let week = 1; week < weeks + 1; week++) {
		const thisWeek = runForWeek(samples, week)
		allSims.push(thisWeek)
	}

	return allSims;
};

//
// const runForRangeOfWeeks = (samples: number[], weeks: number) => {
// 	let allSims = '"95%","80%","50%","20%","5%"\n';
//
// 	for (let week = 1; week < weeks + 1; week++) {
// 		const thisWeek = runForWeek(samples, week);
// 		allSims += `${week},${thisWeek['95']},${thisWeek['80']},${thisWeek['50']},${thisWeek['20']},${thisWeek['5']}
// `;
// 	}
//
// 	return allSims;
// };
//



export {
	runSimulations, runForWeek, runForRangeOfWeeks
};

