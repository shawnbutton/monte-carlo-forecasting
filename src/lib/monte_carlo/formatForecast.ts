import { addDays, addMonths, isValid, lightFormat, parse } from 'date-fns'

const dateToString = (data: Date) => lightFormat(data, 'yyyy-MM-dd')

const addDaysToStart = (startDate: Date, daysToAdd: number) => (periodOn: number) =>
	addDays(startDate, periodOn * daysToAdd)

const addMonthsToStart = (startDate: Date) => (periodOn: number) => addMonths(startDate, periodOn)

const getPeriodAdder = (startDate: Date, period: string) => {
	switch (period) {
		case 'sprint':
			return addDaysToStart(startDate, 14)
		case 'month':
			return addMonthsToStart(startDate)
		default:
			// week
			return addDaysToStart(startDate, 7)
	}
}

export const formatForecast = (
	weeks: number[][],
	startDateString: string,
	period: string,
	percentages: boolean[]
) => {
	let startDate = parse(startDateString, 'yyyy-MM-dd', new Date())
	if (!isValid(startDate)) startDate = new Date()

	const periodAdder = getPeriodAdder(startDate, period)

	const percentValues = percentages
		.map((percentage, i) => {
			return percentage ? (i + 1) * 5 : 0
		})
		.filter(percentage => !!percentage)

	// const header = `"Date", "95%", "80%", "50%", "20%", "5%"\n`
	const header =
		`"Date", ` +
		percentValues
			.map(percentage => {
				return percentage + '%'
			})
			.join(', ') +
		`\n`

	return (
		header +
		weeks
			.map((thisWeek, week) => {
				const thisWeekPercents = percentValues
					.map(percent => {
						return thisWeek[percent]
					})
					.join(', ')

				return `${dateToString(periodAdder(week))}, ` + thisWeekPercents
			})
			.join('\n')
	)
}
