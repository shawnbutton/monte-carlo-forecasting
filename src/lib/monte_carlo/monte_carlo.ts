const runForWeek = (samples: number[], weeks: number, trials: number) => {
	const numSamples = samples.length

	const sims: number[] = []

	for (let i = 0; i < trials; i++) {
		let stories = 0

		for (let week = 0; week < weeks; week++) {
			const index = Math.floor(Math.random() * numSamples)
			stories += samples[index]
		}

		sims.push(stories)
	}

	const sorted = sims.sort((n1, n2) => n1 - n2).reverse()

	const results: number[] = []

	for (let i = 0; i < 99; i++) {
		results[i] = sorted[(i + 1) * (trials / 100)]
	}

	return results
}

const runForRangeOfWeeks = (samples: number[], weeks: number, trials: number) => {
	const allSims: number[][] = []
	for (let week = 1; week < weeks + 1; week++) {
		const thisWeek = runForWeek(samples, week, trials)
		allSims.push(thisWeek)
	}

	return allSims
}

export { runForRangeOfWeeks }
