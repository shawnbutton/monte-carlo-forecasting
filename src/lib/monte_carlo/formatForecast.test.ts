import { describe, expect, it } from 'vitest'
import { formatForecast } from '$lib/monte_carlo/formatForecast'

const arrayFilledWith = (size: number, value: number | boolean) => Array(size).fill(value)

const createMonteCarlo: () => [number[], number[]] = () => [arrayFilledWith(99, 1), arrayFilledWith(99, 2)]

const createPercentages = () => {
	const percentages = arrayFilledWith(19, false)
	percentages[5] = true
	percentages[10] = true
	percentages[15] = true
	return percentages
}

describe('forecast formatter', () => {
	it('formats in weeks', () => {
		const monteCarlo = createMonteCarlo()
		const percentages = createPercentages()
		const result = formatForecast(monteCarlo, '2024-05-05', 'week', percentages)

		expect(result).toMatchSnapshot()
	})

	it('formats in sprints', () => {
		const monteCarlo = createMonteCarlo()
		const percentages = createPercentages()
		const result = formatForecast(monteCarlo, '2024-05-05', 'sprint', percentages)

		expect(result).toMatchSnapshot()
	})

	it('formats in months', () => {
		const monteCarlo = createMonteCarlo()
		const percentages = createPercentages()
		const result = formatForecast(monteCarlo, '2024-05-05', 'months', percentages)

		expect(result).toMatchSnapshot()
	})

	it('formats with a different date', () => {
		const monteCarlo = createMonteCarlo()
		const percentages = createPercentages()
		const result = formatForecast(monteCarlo, '2024-10-10', 'week', percentages)

		expect(result).toMatchSnapshot()
	})

	it('formats with all percentages', () => {
		const monteCarlo = createMonteCarlo()
		const percentages = arrayFilledWith(19, true)
		const result = formatForecast(monteCarlo, '2024-05-05', 'months', percentages)

		expect(result).toMatchSnapshot()
	})

})
