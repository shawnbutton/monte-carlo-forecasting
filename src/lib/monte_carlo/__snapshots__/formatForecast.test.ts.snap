// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`forecast formatter > formats in months 1`] = `
""Date", 30%, 55%, 80%
2024-05-05, 1, 1, 1
2024-05-12, 2, 2, 2"
`;

exports[`forecast formatter > formats in sprints 1`] = `
""Date", 30%, 55%, 80%
2024-05-05, 1, 1, 1
2024-05-19, 2, 2, 2"
`;

exports[`forecast formatter > formats in weeks 1`] = `
""Date", 30%, 55%, 80%
2024-05-05, 1, 1, 1
2024-05-12, 2, 2, 2"
`;

exports[`forecast formatter > formats with a different date 1`] = `
""Date", 30%, 55%, 80%
2024-10-10, 1, 1, 1
2024-10-17, 2, 2, 2"
`;

exports[`forecast formatter > formats with all percentages 1`] = `
""Date", 5%, 10%, 15%, 20%, 25%, 30%, 35%, 40%, 45%, 50%, 55%, 60%, 65%, 70%, 75%, 80%, 85%, 90%, 95%
2024-05-05, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
2024-05-12, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2"
`;
